### PR TITLE
Make `ZoneVariant` a closed, non-exhaustive enum

### PR DESCRIPTION
**Pull request by robertbastian**
#5745

### CODE REVIEW

**Review by Manishearth — APPROVED** [review]
much nicer!

**Review by Manishearth — DISMISSED** [review]

File: components/datetime/src/time_zone.rs [hunk 1]
```diff
@@ -605,12 +605,10 @@ impl FormatTimeZone for SpecificLocationFormat {
...
-        if zone_variant == ZoneVariant::daylight() {
-            &locations.pattern_daylight
-        } else if zone_variant == ZoneVariant::standard() {
-            &locations.pattern_standard
-        } else {
-            &locations.pattern_generic
+        match zone_variant {
+            ZoneVariant::Standard => &locations.pattern_standard,
+            ZoneVariant::Daylight => &locations.pattern_daylight,
+            _ => unreachable!(),
```
Comment (sffc) [inline components/datetime/src/time_zone.rs:611]: Nit: Annotate here and elsewhere that this is unreachable due to the tilde dependency

File: ffi/capi/src/timezone.rs [hunk 2]
```diff
@@ -230,25 +232,29 @@ pub mod ffi {
...
+                Some(icu_timezone::ZoneVariant::Daylight) => "dt",
+                Some(icu_timezone::ZoneVariant::Standard) => "st",
+                _ => return None,
+            });
             Some(())
         }
 
         /// Sets the `zone_variant` field to "standard" time, which may or may
         /// not correspond to a display name with "Standard" in its name.
         #[diplomat::rust_link(icu::timezone::ZoneVariant::standard, FnInStruct)]
```
Comment (sffc) [inline ffi/capi/src/timezone.rs:248]: Nit: Update the rust link

File: ffi/capi/src/timezone.rs [hunk 3]
```diff
@@ -211,17 +211,19 @@ pub mod ffi {
         /// Sets the `zone_variant` field from a string.
         ///
         /// Returns null if the string is not a valid zone variant.
-        #[diplomat::rust_link(icu::timezone::ZoneVariant, Struct, compact)]
+        #[diplomat::rust_link(icu::timezone::ZoneVariant, Enum, compact)]
         #[diplomat::rust_link(icu::timezone::ZoneVariant::from_str, FnInStruct, hidden)]
         pub fn try_set_zone_variant(&mut self, id: &DiplomatStr) -> Option<()> {
```
Comment (sffc) [inline ffi/capi/src/timezone.rs:216]: Nit: Delete these old functions. New variants will get new functions.

**Review by sffc — APPROVED** [review]

**Pull request merged**